File: app/src/main/java/com/esorokin/lantector/ui/activity/MainActivity.java
package com.esorokin.lantector.ui.activity;

import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.NavigationView;
import android.support.v4.view.GravityCompat;
import android.support.v7.app.ActionBarDrawerToggle;
import android.view.MenuItem;

import com.arellomobile.mvp.presenter.InjectPresenter;
import com.esorokin.lantector.R;
import com.esorokin.lantector.databinding.ActivityMainBinding;
import com.esorokin.lantector.presentation.navigation.AppScreen;
import com.esorokin.lantector.presentation.navigation.TabsInfo;
import com.esorokin.lantector.presentation.presenter.MainPresenter;
import com.esorokin.lantector.presentation.view.MainView;
import com.esorokin.lantector.ui.plugins.SwitchFragmentNavigationPlugin;
import com.esorokin.lantector.ui.plugins.ToolbarPlugin;

public class MainActivity extends BaseActivity implements MainView, NavigationView.OnNavigationItemSelectedListener, SwitchFragmentNavigationPlugin.ScreenSwitchListener {
	@InjectPresenter
	MainPresenter presenter;

	private TabsInfo tabsInfo = new TabsInfo();
	private ActivityMainBinding viewBinding;

	@Override
	protected void initPlugins() {
		super.initPlugins();
		compositionPlugin().attach(new ToolbarPlugin(this));
		compositionPlugin().attach(new SwitchFragmentNavigationPlugin(this, R.id.container, tabsInfo, this));
	}

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		viewBinding = DataBindingUtil.setContentView(this, R.layout.activity_main);
		initTabsInfo();

		ActionBarDrawerToggle drawerToggle = new ActionBarDrawerToggle(this, viewBinding.drawerLayout, viewBinding.header.toolbar, R.string.drawer_open, R.string.drawer_close);

		viewBinding.drawerLayout.addDrawerListener(drawerToggle);
		drawerToggle.syncState();

		viewBinding.navigationView.setNavigationItemSelectedListener(this);
	}

	private void initTabsInfo() {
		tabsInfo.add(DETECT_LANGUAGE_SCREEN_NAME, R.id.detectLanguageTab, new AppScreen.DetectLanguage());
		tabsInfo.add(HISTORY_SCREEN_NAME, R.id.historyTab, new AppScreen.History());
	}

	@Override
	public void onScreenSwitched(String screenName) {
		viewBinding.navigationView.getMenu().findItem(tabsInfo.getTabId(screenName)).setChecked(true);
	}

	@Override
	public void onBackPressed() {
		if (viewBinding.drawerLayout.isDrawerOpen(GravityCompat.START)) {
			viewBinding.drawerLayout.closeDrawer(GravityCompat.START);
		} else {
			super.onBackPressed();
		}
	}

	@Override
	public boolean onNavigationItemSelected(@NonNull MenuItem item) {
		presenter.userSelectTab(tabsInfo.getScreenName(item.getItemId()));
		return true;
	}

	@Override
	public void openDrawer() {
		viewBinding.drawerLayout.openDrawer(GravityCompat.START);
	}

	@Override
	public void closeDrawer() {
		viewBinding.drawerLayout.closeDrawer(GravityCompat.START);
	}
}

File: app/src/main/java/com/esorokin/lantector/ui/activity/MainActivity.kt
package com.esorokin.lantector.ui.activity

import android.os.Bundle
import android.support.design.widget.NavigationView
import android.support.v4.view.GravityCompat
import android.support.v7.app.ActionBarDrawerToggle
import android.view.MenuItem
import com.arellomobile.mvp.presenter.InjectPresenter
import com.esorokin.lantector.R
import com.esorokin.lantector.presentation.navigation.AppScreen
import com.esorokin.lantector.presentation.navigation.TabsInfo
import com.esorokin.lantector.presentation.presenter.MainPresenter
import com.esorokin.lantector.presentation.view.MainView
import com.esorokin.lantector.presentation.view.MainView.TabScreens.DETECT_LANGUAGE_SCREEN_NAME
import com.esorokin.lantector.presentation.view.MainView.TabScreens.HISTORY_SCREEN_NAME
import com.esorokin.lantector.ui.plugin.SwitchFragmentNavigationPlugin
import com.esorokin.lantector.ui.plugin.ToolbarPlugin
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.include_default_simple_toolbar.*

class MainActivity : BaseActivity(), MainView, NavigationView.OnNavigationItemSelectedListener {
    @InjectPresenter
    internal lateinit var presenter: MainPresenter

    private lateinit var toolbarPlugin: ToolbarPlugin
    private lateinit var fragmentNavigationPlugin: SwitchFragmentNavigationPlugin

    private val tabsInfo: TabsInfo = TabsInfo().apply {
        add(DETECT_LANGUAGE_SCREEN_NAME, R.id.detectLanguageTab, AppScreen.DetectLanguage())
        add(HISTORY_SCREEN_NAME, R.id.historyTab, AppScreen.History())
    }

    override fun initPlugins() {
        super.initPlugins()
        toolbarPlugin = compositionPlugin.attach(ToolbarPlugin({ this }))
        fragmentNavigationPlugin = compositionPlugin.attach(SwitchFragmentNavigationPlugin(this, R.id.container, tabsInfo, onScreenSwitched()))
    }

    private fun onScreenSwitched(): (String) -> Unit = {
        navigationView.menu.findItem(tabsInfo.getTabId(it)).isChecked = true
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        ActionBarDrawerToggle(this, drawerLayout, toolbar, R.string.drawer_open, R.string.drawer_close).also {
            drawerLayout.addDrawerListener(it)
            it.syncState()
        }

        navigationView.setNavigationItemSelectedListener(this)
    }

    override fun onBackPressed() {
        if (drawerLayout.isDrawerOpen(GravityCompat.START)) {
            drawerLayout.closeDrawer(GravityCompat.START)
        } else {
            super.onBackPressed()
        }
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        presenter.userSelectTab(tabsInfo.getScreenName(item.itemId))
        return true
    }

    override fun openDrawer() {
        drawerLayout.openDrawer(GravityCompat.START)
    }

    override fun closeDrawer() {
        drawerLayout.closeDrawer(GravityCompat.START)
    }
}
